zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/SpotRestorationTool.java

```diff
@@ -0,0 +1,284 @@
+import org.apache.zookeeper.server.DataTree;
+import org.apache.zookeeper.server.persistence.FileTxnSnapLog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a tool for automatic restoration of ZK data based on a zk node path using a collection of
+ * snapshot and transaction logs.
+ */
+public class SpotRestorationTool {
+  private static final Logger LOG = LoggerFactory.getLogger(SpotRestorationTool.class);
+
+  private ZooKeeper zk;
+  private FileTxnSnapLog snapLog;
+  private boolean restoreRecursively;
+  private String targetZNodePath;
+  // A list storing messages about skipped nodes, which will be shown to the user at the end of spot restoration
+  private List<String> messages;
+
+  /**
+   * Constructor
+   * @param dataDir The directory contains fully-restored snapshot and transaction log files to be used for spot restoration
+   * @param zk A client connection to the zk server to be restored
+   * @param targetZNodePath The znode path to restore
+   * @param restoreRecursively If true then restore the whole sub-data tree of the target znode;
+   *                           if false then restore the target znode only
+   * @throws IOException
+   */
+  public SpotRestorationTool(File dataDir, ZooKeeper zk, String targetZNodePath,
+      boolean restoreRecursively) throws IOException {
+    if (dataDir == null || !dataDir.exists()) {
+      throw new IllegalArgumentException(
+          "The provided dataDir is null or does not exist, please check the input.");
+    }
+    this.zk = zk;
+    this.snapLog = new FileTxnSnapLog(dataDir, dataDir);
+    this.targetZNodePath = targetZNodePath;
+    this.restoreRecursively = restoreRecursively;
+    this.messages = Lists.newArrayList();
+  }
+
+  /**
+   * Run the spot restoration.
+   * 1. If a node exists in restored data tree but not in the zk server, create the node in the server
+   * 2. If a node exists in zk server but not the restored data tree, show messages that inform user
+   *  to manually delete the node after this spot restoration run
+   * 3. If a node exists in both zk server and restored data tree, show message that inform user to
+   *  manually delete the node after this spot restoration run and run spot restoration on the node path again
+   * @throws IOException
+   */
+  public void run() throws IOException {
+    LOG.info(
+        "Starting spot restoration for znode path " + targetZNodePath + ", using data provided in "
+            + snapLog.getDataDir().getPath());
+    DataTree dataTree = new DataTree();
+    long zxidRestored = snapLog.restore(dataTree, new ConcurrentHashMap<>(), (hdr, rec, digest) -> {
+      // Do nothing since we are trying to build a data tree in memory, not in actual zk server
+    });
+
+    DataNode originalNode = dataTree.getNode(targetZNodePath);
+    if (originalNode == null) {
+      return;
+    }
+    Stat originalNodeStat = new Stat();
+    originalNode.copyStat(originalNodeStat);
```

Comment:
I believe the name we want here is `restored`, not `original`?

Reply:
I used `original` because it is the target znode which the user specified by passing in the znode path. But I agree this word can be confusing, let me change it to `targetRestoredNode`.